Feature/unit test source builds

tleyden opened the pull request:
#### Fixes #.

- [x] No need to run `flake8` 
- [x] No need to run `run_repo_tests.sh`

#### Changes proposed in this pull request:

- Run sg unit tests on source builds

#### Passing functional tests

This got past the build/test phase: 

http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-topology-specific-di/278/console

Logs:

```
TASK [SG ACCEL | Run bootstrap.sh script] **************************************
skipping: [sg1]
skipping: [sg2]
changed: [ac1]

TASK [SYNC GATEWAY & SG ACCEL | Build] *****************************************
changed: [sg1]
changed: [sg2]
changed: [ac1]

TASK [SYNC GATEWAY & SG ACCEL | Unit Tests] ************************************
changed: [sg2]
changed: [sg1]
changed: [ac1]
```

tleyden commented:
Not sure why https://github.com/couchbaselabs/mobile-testkit/pull/1100/commits/ea48a3b6bbf2c47d4e4bf8c1087e9cde0ca3da51 didn't work -- but I just reverted it since it didn't add any value

sethrosetter commented:
@tleyden Can you run this against a build you would expect this to fail on just to make sure that ansible will fail when unit tests fail?

tleyden commented:
Good call

tleyden commented:
Looks like it's correctly detecting the unit test failures:

http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-topology-specific-di/280/consoleFull

```
TASK [SYNC GATEWAY & SG ACCEL | Unit Tests] ************************************
changed: [sg1]
changed: [sg2]
fatal: [ac1]: FAILED! => {"changed": true, "cmd": "./test.sh", "delta": "0:03:07.401956", "end": "2017-04-04 15:44:26.382277", "failed": true, "rc": 1, "start": "2017-04-04 15:41:18.980321", "stderr": "", "stdout": "Testing code with 'go test' ...\nRunning Sync Gateway unit tests\n?   \tgithub.com/couchbase/sync_gateway\t[no test files]\n2017-04-04T15:41:21.394-07:00 Opening Walrus database sync_gateway_tests on <walrus:>\n=== RUN   TestValidateGuestUser\n--- PASS: TestValidateGuestUser (0.00s)\n=== RUN   TestValidateUser\n--- PASS: TestValidateUser (0.09s)\n=== RUN   TestValidateRole\n--- PASS: TestValidateRole (0.00s)\n=== RUN   TestValidateUserEmail\n--- PASS: TestValidateUserEmail (0.09s)\n=== RUN   TestUserPasswords\n--- PASS: TestUserPasswords (0.78s)\n=== RUN   TestAuthenticationSpeed\n--- PASS: TestAuthenticationSpeed (0.18s)\n=== RUN   TestSerializeUser\n2017-04-04 15:41:22.636707 I | Marshaled User as: {\"name\":\"me\",\"admin_channels\":{\"me\":1,\"public\":1},\"all_channels\":
```

sethrosetter commented:
Why is the job not marked as failed?

tleyden commented:
I believe it's going to run through all of the tests and mark all tests where the unit tests as failed as being failed.  Maybe we should wait until the test finishes.

I'm not sure how to make it fail earlier .. maybe `test.sh` isn't returning the correct exit code.  I'll look into it

sethrosetter commented:
Since `test.sh` (I think) is happening during initial provisioning, I would expect the provisioning to raise an exception and fail all the tests quickly. If it is running the functional tests, then this probably was not caught where it should have.

tleyden commented:
I wonder if it's due to `STREAM_PYTEST_OUTPUT` -- disabling and [re-running](http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-topology-specific-di/281/console).

`test.sh` looks like it's fine since it's using `set -e` which will abort as soon as it hits a non-zero exit code.  When I ran locally and got the same test failure:

```
echo $?
1
```

tleyden commented:
Here's a completed job where the unit tests failed:

http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-topology-specific-di/281/

I think that's a lot better than the current behavior, where it passes all the functional tests.

I have no idea how to fix the "early failure", maybe you should spin off a separate ticket .. from what I can tell `test.sh` and ansible are behaving correctly.

sethrosetter commented:
@tleyden This behavior looks expected. Thanks!
